README.md
# Cylon.js API plugin for HTTP

Cylon.js (http://cylonjs.com) is a JavaScript framework for robotics, physical computing, and the Internet of Things (IoT).

API plugins are separate from the Cylon.js main module, to allow for greater modularization of Cylon robots.

This repository contains the Cylon API plugin for HTTP.

For more information about Cylon, check out the repo at https://github.com/hybridgroup/cylon

[![Build Status](https://travis-ci.org/hybridgroup/cylon-api-socketio.svg)](https://travis-ci.org/hybridgroup/cylon-api-socketio)
[![Code Climate](https://codeclimate.com/github/hybridgroup/cylon-api-socketio/badges/gpa.svg)](https://codeclimate.com/github/hybridgroup/cylon-api-socketio)
[![Test Coverage](https://codeclimate.com/github/hybridgroup/cylon-api-socketio/badges/coverage.svg)](https://codeclimate.com/github/hybridgroup/cylon-api-socketio)

## How To Install

    $ npm install cylon cylon-api-http

## How To Use

With both of these modules installed, you can make your robots accessible over HTTP:

```javascript
"use strict";

var Cylon = require("cylon");

Cylon.api("http", {
  ssl: false // serve unsecured, over HTTP

  // optional configuration here.
  // for details see 'Configuration' section.
});

Cylon.robot({
  name: "Maria",

  connections: {
    arduino: { adaptor: 'firmata', port: '/dev/ttyACM0' }
  },

  devices: {
    led: { driver: 'led', pin: 13 }
  },

  work: function() {
    // we'll interact with this robot through the API
  }
}).start();
```

## How To Connect

Once you've got a Cylon instance with the API configured and running, you can use cURL to test it out:

      $ curl http://127.0.0.1:4000/api/robots/Maria
      { "robot": { "name": "Maria" /* ... */ } }

## Routes

This plugin follows the [CPPP-IO][] specification.

To get started, you can visit the `/api` route in your browser to get an overview of the Cylon Master Control Program (MCP).
This provides a high level overview of your program:

      $ curl http://127.0.0.1:4000/api | python -m json.tool
      {
          "MCP": {
              "commands": [ ],
              "events": [ ],
              "robots": [
                  {
                      "name": "Maria",
                      "connections": [
                          {
                              "name": "arduino",
                              "adaptor": "firmata",
                              "details": { "port": "/dev/ttyACM0" }
                          }
                      ],
                      "devices": [
                          {
                              "name": "led",
                              "commands": [ "toggle" ],
                              "connection": "arduino",
                              "details": { "pin": "13" },
                              "driver": "LED"
                          }
                      ],
                      "commands": [ ],
                      "events": [ ]
                  }
              ]
          }
      }

For details on other available routes in the API and expected responses, please see the [specification document][spec].

[CPPP-IO]: https://github.com/hybridgroup/cppp-io

## Robeaux

This plugin includes a default front-end in [Robeaux][].

Robeaux is a simple tool to use against the HTTP API, and explore your robots visually, while also able to issue commands to robots and listen for events.

## Configuration

Option     | Description
------     | -----------
`host`     | What HTTP host to serve from. Defaults to `127.0.0.1`.
`port`     | What HTTP port to serve on. Defaults to `3000`.
`auth`     | What authorization scheme to use. e.g. `{ type: "basic", user: "username", pass: "password"}`. Defaults to `false`.
`CORS`     | Cross-Origin Resource Sharing option (HTTP Header: "Access-Control-Allow-Origin"). Defaults to "*".
`serveDir` | Directory to serve as static assets. Defaults to the included [Robeaux][] installation.
`ssl`      | SSL `key` and `cert` options, wrapped as an object. Set to `false` for an unsecured API. Defaults to included self-signed certs.

[Robeaux]: https://github.com/hybridgroup/robeaux

## Documentation

We're busy adding documentation to [cylonjs.com](http://cylonjs.com). Please check there as we continue to work on Cylon.js.

Thank you!

## Contributing

For our contribution guidelines, please go to [https://github.com/hybridgroup/cylon/blob/master/CONTRIBUTING.md](https://github.com/hybridgroup/cylon/blob/master/CONTRIBUTING.md).

## Release History

For the release history, please go to [https://github.com/hybridgroup/cylon-api-http/blob/master/RELEASES.md](https://github.com/hybridgroup/cylon-api-http/blob/master/RELEASES.md).

## License

Copyright (c) 2014-2016 The Hybrid Group. Licensed under the Apache 2.0 license.


RELEASES.md
## Release History

0.4.1 - Handle new install location for npm 3.x

0.4.0 - Update dependencies, add newer Robeaux version

0.3.0 - Update dependencies, add newer Robeaux version

0.2.0 - Continued internals, fixes, `serveDir` option

0.1.1 - Correct command issues, re-organize internals.

0.1.0 - Initial release


spec/helper.js
"use strict";

process.env.NODE_ENV = "test";

var path = require("path");

var chai = require("chai"),
    sinon = require("sinon"),
    sinonChai = require("sinon-chai");

chai.use(sinonChai);

global.chai = chai;
global.sinon = sinon;

global.should = chai.should();
global.expect = chai.expect;
global.assert = chai.assert;
global.AssertionError = chai.AssertionError;

global.spy = sinon.spy;
global.stub = sinon.stub;

// convenience function to require modules in lib directory
global.lib = function(module) {
  return require(path.normalize("./../lib/" + module));
};

global.support = function(module) {
  return require("./support/" + module);
};


spec/lib/api.spec.js
"use strict";

var https = require("https"),
    path = require("path");

var API = lib("api");

var MockRequest = support("mock_request"),
    MockResponse = support("mock_response");

describe("API", function() {
  var api;

  beforeEach(function() {
    api = new API();
    stub(console, "log");
  });

  afterEach(function() {
    console.log.restore();
  });

  describe("constructor", function() {
    var mod;

    beforeEach(function() {
      mod = new API({
        host: "0.0.0.0",
        port: "1234"
      });
    });

    it("sets @express to an Express instance", function() {
      expect(api.express.listen).to.be.a("function");
    });

    it("sets default values", function() {
      expect(api.host).to.be.eql("127.0.0.1");
      expect(api.port).to.be.eql("3000");
    });

    it("overrides default values if passed to constructor", function() {
      expect(mod.host).to.be.eql("0.0.0.0");
      expect(mod.port).to.be.eql("1234");
    });
  });

  describe("default", function() {
    it("host", function() {
      expect(api.host).to.be.eql("127.0.0.1");
    });

    it("port", function() {
      expect(api.port).to.be.eql("3000");
    });

    it("auth", function() {
      expect(api.auth).to.be.eql(false);
    });

    it("CORS", function() {
      expect(api.CORS).to.be.eql("");
    });

    it("ssl", function() {
      var sslDir = path.join(__dirname, "/../../ssl/");
      expect(api.ssl.key).to.be.eql(sslDir + "server.key");
      expect(api.ssl.cert).to.be.eql(sslDir + "server.crt");
    });
  });

  describe("#createServer", function() {
    it("sets @express to an express server", function() {
      api.express = null;
      api.createServer();
      expect(api.express).to.be.a("function");
    });

    context("if SSL is configured", function() {
      it("sets @server to a https.Server instance", function() {
        api.createServer();
        expect(api.server).to.be.an.instanceOf(https.Server);
      });
    });

    context("if SSL is not configured", function() {
      beforeEach(function() {
        api.ssl = false;
        api.createServer();
      });

      it("logs that the API is insecure", function() {
        expect(console.log).to.be.calledWithMatch("insecure connection");
      });

      it("sets @server to @express", function() {
        expect(api.server).to.be.eql(api.express);
      });
    });
  });

  describe("#setupAuth", function() {
    context("when auth.type is basic", function() {
      beforeEach(function() {
        api.auth = { type: "basic", user: "user", pass: "pass" };
      });

      it("returns a basic auth middleware function", function() {
        var fn = api.setupAuth(),
            req = new MockRequest(),
            res = new MockResponse(),
            next = spy();

        var auth = "Basic " + new Buffer("user:pass").toString("base64");

        req.headers.authorization = auth;

        fn(req, res, next);
        expect(next).to.be.called;
      });
    });

    context("when auth is null", function() {
      beforeEach(function() {
        api.auth = null;
      });

      it("returns a pass-through middleware function", function() {
        var fn = api.setupAuth(),
            next = spy();

        fn(null, null, next);
        expect(next).to.be.called;
      });
    });
  });

  describe("#listen", function() {
    beforeEach(function() {
      // we create a plain HTTP server to avoid a log message from Node
      api.ssl = false;
      api.createServer();

      stub(api.server, "listen").yields();

      api.listen();
    });

    afterEach(function() {
      api.server.listen.restore();
    });

    it("listens on the configured host and port", function() {
      expect(api.server.listen).to.be.calledWith("3000", "127.0.0.1");
    });

    context("when the server is running", function() {
      it("logs that it's online and listening", function() {
        expect(console.log).to.be.calledWith(
          "Cylon HTTP API server is now online."
        );

        expect(console.log).to.be.calledWith(
          "Listening at %s://%s:%s", "http", "127.0.0.1", "3000"
        );
      });
    });
  });
});


spec/lib/routes.spec.js
"use strict";

var _ = require("lodash");

var router = lib("routes");

var MockRequest = support("mock_request"),
    MockResponse = support("mock_response"),
    MockPromise = support("mock_promise");

function findRoute(path) {
  return _.find(router.stack, function(route) {
    return route.regexp.test(path);
  });
}

function findFinalHandler(path) {
  return _(findRoute(path).route.stack).map("handle").last();
}

describe("API routes", function() {
  var routes = [
    ["GET", "/"],
    ["GET", "/events"],
    ["GET", "/events/event"],
    ["GET", "/commands"],
    ["POST", "/commands/command"],
    ["GET", "/robots"],
    ["GET", "/robots/TestBot"],
    ["GET", "/robots/TestBot/commands"],
    ["POST", "/robots/TestBot/commands/hello"],
    ["GET", "/robots/TestBot/events"],
    ["POST", "/robots/TestBot/events/hello"],
    ["GET", "/robots/TestBot/devices"],
    ["GET", "/robots/TestBot/devices/ping"],
    ["GET", "/robots/TestBot/devices/ping/events"],
    ["GET", "/robots/TestBot/devices/ping/events/ping"],
    ["GET", "/robots/TestBot/devices/ping/commands"],
    ["POST", "/robots/TestBot/devices/ping/commands/ping"],
    ["GET", "/robots/TestBot/connections"],
    ["GET", "/robots/TestBot/connections/loopback"]
  ];

  routes.forEach(function(route) {
    var method = route[0],
        path = route[1];

    it("defines a " + method + " route for " + path, function() {
      expect(findRoute(path)).to.be.a("object");
    });
  });
});

describe("API commands", function() {
  var req, res;

  beforeEach(function() {
    req = new MockRequest();
    res = new MockResponse();

    req.mcp = {
      commands: {
        ping: function() { return "pong"; },
        pingAsync: function() {
          var promise = new MockPromise();
          return promise.resolve("immediate pong");
        }
      }
    };

    res.status = function(statusCode) {
      res.statusCode = statusCode;
      return res;
    };

    req.device = {
      name: "testDevice",
      commands: {
        announce: function() { return "im here"; },
        announceAsync: function() {
          var promise = new MockPromise();

          process.nextTick(function() {
            return promise.reject("sorry, sore throat");
          });

          return promise.deferred;
        }
      }
    };

    req.robot = {
      name: "fred",

      commands: {
        speak: function() { return "ahem"; },
        speakAsync: function() {
          var promise = new MockPromise();

          process.nextTick(function() {
            return promise.resolve("see ya in another cycle");
          });

          return promise.deferred;
        }
      },

      devices: {
        testDevice: req.device
      }
    };
  });

  it("returns the list of MCP commands", function() {
    res.json = function(obj) {
      expect(obj.commands).to.be.an("array");
      expect(obj.commands.length).to.equal(2);
      expect(obj.commands[0]).to.equal("ping");
    };

    findFinalHandler("/commands")(req, res);
  });

  it("invokes an MCP command", function() {
    req.params = { command: "ping" };

    res.json = function(obj) {
      expect(obj.result).to.equal("pong");
    };

    findFinalHandler("/commands/ping")(req, res);
  });

  it("returns an immediate MCP async command", function() {
    req.params = { command: "pingAsync" };

    res.json = function(obj) {
      expect(obj.result).to.equal("immediate pong");
    };

    findFinalHandler("/commands/pingAsync")(req, res);
  });

  it("returns the list of robot commands", function() {
    req.params = {robot: "fred"};

    res.json = function(obj) {
      expect(obj.commands).to.be.an("array");
      expect(obj.commands.length).to.equal(2);
      expect(obj.commands[0]).to.equal("speak");
    };

    findFinalHandler("/robots/fred/commands")(req, res);
  });

  it("invokes a robot command", function() {
    req.params = { robot: "fred", command: "speak" };

    res.json = function(obj) {
      expect(obj.result).to.equal("ahem");
    };

    findFinalHandler("/robots/fred/commands/speak")(req, res);
  });

  it("invokes an asynchronous robot command", function() {
    req.params = { robot: "fred", command: "speakAsync" };

    res.json = function(obj) {
      expect(obj.result).to.equal("see ya in another cycle");
    };

    findFinalHandler("/robots/fred/commands/speakAsync")(req, res);
  });

  it("returns the list of device commands", function() {
    req.params = { robot: "fred", device: "testDevice" };

    res.json = function(obj) {
      expect(obj.commands).to.be.an("array");
      expect(obj.commands.length).to.equal(2);
      expect(obj.commands[0]).to.equal("announce");
    };

    var path = "/robots/fred/devices/testDevice/commands";
    findFinalHandler(path)(req, res);
  });

  it("invokes a device command", function() {
    req.params = { robot: "fred", device: "testDevice", command: "announce" };

    res.json = function(obj) {
      expect(obj.result).to.equal("im here");
    };

    var path = "/robots/fred/devices/testDevice/commands/announce";
    findFinalHandler(path)(req, res);
  });

  it("returns correctly for an async device command that errored", function() {
    req.params = {
      robot: "fred",
      device: "testDevice",
      command: "announceAsync"
    };

    res.json = function(obj) {
      expect(obj.error).to.equal("sorry, sore throat");
      expect(res.statusCode).to.equal(500);
    };

    var path = "/robots/fred/devices/testDevice/commands/announceAsync";
    findFinalHandler(path)(req, res);
  });
});


lib/routes.js
/*
 * Cylon API - Route Definitions
 * cylonjs.com
 *
 * Copyright (c) 2013-2015 The Hybrid Group
 * Licensed under the Apache 2.0 license.
*/

"use strict";

/* eslint new-cap: 0, quote-props: 0, max-len: 0 */

var router = module.exports = require("express").Router();

var get = router.get.bind(router),
    post = router.post.bind(router),
    all = router.all.bind(router);

var loader = require("./middleware/loader");

var eventHeaders = {
  "Content-Type": "text/event-stream",
  "Connection": "keep-alive",
  "Cache-Control": "no-cache"
};

// Run an MCP, Robot, or Device command.  Process the results immediately,
// or asynchronously if the result is a Promise.
function runCommand(req, res, my, command) {
  var result = command.apply(my, req.commandParams),
      promise = typeof result === "object" && typeof result.then === "function";

  function respond(r) { res.json({ result: r }); }
  function error(err) { res.status(500).json({ error: err }); }

  return promise ? result.then(respond).catch(error) : respond(result);
}

get("/", function(req, res) {
  res.json({ MCP: req.mcp.toJSON() });
});

get("/events", function(req, res) {
  res.json({ events: req.mcp.events });
});

get("/events/:event", function(req, res) {
  var event = req.params.event;

  res.writeHead(200, eventHeaders);

  var writeData = function(data) {
    res.write("data: " + JSON.stringify(data) + "\n\n");
  };

  req.mcp.on(event, writeData);

  res.on("close", function() {
    req.mcp.removeListener(event, writeData);
  });
});

get("/commands", function(req, res) {
  res.json({ commands: Object.keys(req.mcp.commands) });
});

post("/commands/:command", function(req, res) {
  var command = req.mcp.commands[req.params.command];
  runCommand(req, res, req.mcp, command);
});

get("/robots", function(req, res) {
  res.json({ robots: req.mcp.toJSON().robots });
});

get("/robots/:robot", loader, function(req, res) {
  res.json({ robot: req.robot });
});

get("/robots/:robot/commands", loader, function(req, res) {
  res.json({ commands: Object.keys(req.robot.commands) });
});

all("/robots/:robot/commands/:command", loader, function(req, res) {
  var command = req.robot.commands[req.params.command];
  runCommand(req, res, req.robot, command);
});

get("/robots/:robot/events", loader, function(req, res) {
  res.json({ events: req.robot.events });
});

all("/robots/:robot/events/:event", loader, function(req, res) {
  var event = req.params.event;

  res.writeHead(200, eventHeaders);

  var writeData = function(data) {
    res.write("data: " + JSON.stringify(data) + "\n\n");
  };

  req.robot.on(event, writeData);

  res.on("close", function() {
    req.robot.removeListener(event, writeData);
  });
});

get("/robots/:robot/devices", loader, function(req, res) {
  res.json({ devices: req.robot.toJSON().devices });
});

get("/robots/:robot/devices/:device", loader, function(req, res) {
  res.json({ device: req.device });
});

get("/robots/:robot/devices/:device/events", loader, function(req, res) {
  res.json({ events: req.device.events });
});

get("/robots/:robot/devices/:device/events/:event", loader, function(req, res) {
  var event = req.params.event;

  res.writeHead(200, eventHeaders);

  function writeData(data) {
    res.write("data: " + JSON.stringify(data) + "\n\n");
  }

  req.device.on(event, writeData);

  res.on("close", function() {
    req.device.removeListener(event, writeData);
  });
});

get("/robots/:robot/devices/:device/commands", loader, function(req, res) {
  res.json({ commands: Object.keys(req.device.commands) });
});

all("/robots/:robot/devices/:device/commands/:command", loader, function(req, res) {
  var command = req.device.commands[req.params.command];
  runCommand(req, res, req.device, command);
});

get("/robots/:robot/connections", loader, function(req, res) {
  res.json({ connections: req.robot.toJSON().connections });
});

get("/robots/:robot/connections/:connection", loader, function(req, res) {
  res.json({ connection: req.connection });
});


lib/auth/basic.js
/*
 * Cylon API - Basic Auth
 * cylonjs.com
 *
 * Copyright (c) 2013-2015 The Hybrid Group
 * Licensed under the Apache 2.0 license.
*/

"use strict";

var http = require("http");

var unauthorized = function unauthorized(res) {
  res.statusCode = 401;
  res.setHeader("WWW-Authenticate", "Basic realm=\"Authorization Required\"");
  res.end("Unauthorized");
};

var error = function error(code, msg) {
  var err = new Error(msg || http.STATUS_CODES[code]);
  err.status = code;
  return err;
};

module.exports = function(config) {
  var user = config.user,
      pass = config.pass;

  return function auth(req, res, next) {
    var authorization = req.headers.authorization;

    if (!authorization) {
      return unauthorized(res);
    }

    // malformed
    var parts = authorization.split(" ");

    if (parts[0].toLowerCase() !== "basic" || !parts[1]) {
      return next(error(400));
    }

    authorization = parts[1];

    // credentials
    authorization = new Buffer(authorization, "base64").toString();
    authorization = authorization.match(/^([^:]+):(.+)$/);

    if (!authorization) {
      return unauthorized(res);
    }

    if (authorization[1] === user && authorization[2] === pass) {
      return next();
    }

    return unauthorized(res);
  };
};


lib/middleware/loader.js
/*
 * Cylon HTTP API - Loader Middleware
 * cylonjs.com
 *
 * Copyright (c) 2013-2015 The Hybrid Group
 * Licensed under the Apache 2.0 license.
*/

"use strict";

// loads robot/device/connection as appropriate for the current request
module.exports = function loader(req, res, next) {
  var robot = req.params.robot,
      device = req.params.device,
      connection = req.params.connection;

  if (robot) {
    req.robot = req.mcp.robots[robot];
    if (!req.robot) {
      return res.status(404).json({
        error: "No robot found with the name " + robot
      });
    }
  }

  if (robot && device) {
    req.device = req.robot.devices[device];
    if (!req.device) {
      return res.status(404).json({
        error: "No device found with the name " + device
      });
    }
  }

  if (robot && connection) {
    req.connection = req.robot.connections[connection];
    if (!req.connection) {
      return res.status(404).json({
        error: "No connection found with the name " + connection
      });
    }
  }

  next();
};


lib/api.js
/*
 * Cylon API
 * cylonjs.com
 *
 * Copyright (c) 2013-2015 The Hybrid Group
 * Licensed under the Apache 2.0 license.
*/

/* eslint no-sync: 0 */

"use strict";

var fs = require("fs"),
    path = require("path");

var express = require("express"),
    bodyParser = require("body-parser"),
    _ = require("lodash");

var defaults = {
  host: "127.0.0.1",
  port: "3000",
  auth: false,
  CORS: "",
  serveDir: path.join(__dirname, "/../node_modules/robeaux/"),
  ssl: {
    key: path.join(__dirname, "/../ssl/server.key"),
    cert: path.join(__dirname, "/../ssl/server.crt")
  }
};

// handle the flattened location for dependencies using npm 3.x
if (!fs.existsSync(defaults.serveDir)) {
  defaults.serveDir = path.join(__dirname, "/../../robeaux");
}

var API = module.exports = function API(opts) {
  opts = opts || {};

  this.mcp = opts.mcp;

  _.forEach(defaults, function(def, name) {
    this[name] = _.has(opts, name) ? opts[name] : def;
  }, this);

  this.createServer();
  this.configureMiddleware();
  this.loadRoutes();

  // error handling
  this.express.use(function(err, req, res, next) {
    res.status(500).json({ error: err.message || "An error occured."});
    next();
  });
};

API.prototype.createServer = function createServer() {
  this.express = express();

  // configure ssl if requested
  if (this.ssl && typeof this.ssl === "object") {
    var https = require("https");

    this.server = https.createServer({
      key: fs.readFileSync(this.ssl.key),
      cert: fs.readFileSync(this.ssl.cert)
    }, this.express);
  } else {
    console.log("API using insecure connection.");
    console.log("We recommend using an SSL certificate with Cylon.");
    this.server = this.express;
  }
};

API.prototype.configureMiddleware = function configureMiddleware() {
  // auth
  this.express.use(this.setupAuth());

  // parse request bodies
  this.express.use(bodyParser.json());
  this.express.use(bodyParser.urlencoded({ extended: true }));

  // serve static content
  this.express.use(express.static(this.serveDir));

  // set CORS headers for API requests
  this.express.use(function(req, res, next) {
    res.set("Access-Control-Allow-Origin", this.CORS || "*");
    res.set("Access-Control-Allow-Headers", "Content-Type");
    res.set("Content-Type", "application/json");

    if (req.method === "OPTIONS") {
      return res.sendStatus(200);
    }

    return next();
  }.bind(this));

  // extracts command params from request
  this.express.use(function(req, res, next) {
    req.commandParams = _.values(req.body);
    return next();
  });

  // set Cylon/MCP in request
  this.express.use(function(req, res, next) {
    req.mcp = this.mcp;
    return next();
  }.bind(this));
};

API.prototype.loadRoutes = function loadRoutes() {
  this.express.use("/api", require("./routes"));
};

API.prototype.setupAuth = function setupAuth() {
  var authfn = function auth(req, res, next) { next(); };

  if (!this.auth || !this.auth.type) { return authfn; }

  var type = this.auth.type,
      module = "./auth/" + type,
      filename = path.join(__dirname, module + ".js"),
      exists = fs.existsSync(filename);

  if (exists) {
    authfn = require(filename)(this.auth);
  }

  return authfn;
};

API.prototype.start = API.prototype.listen = function() {
  var protocol = this.ssl ? "https" : "http",
      host = this.host,
      port = this.port;

  this.server.listen(this.port, this.host, null, function() {
    console.log("Cylon HTTP API server is now online.");
    console.log("Listening at %s://%s:%s", protocol, host, port);
  });
};
